### PR TITLE
feat(i18n): add `Astro.currentLocale`

### PR DESCRIPTION
**Pull request by ematipico**
## Changes

RFC change: https://github.com/withastro/roadmap/pull/734/commits/f934880ab5835492e56ddbb69e3f32662310c8bf

This PR adds a new `Astro.currentLocale`, which is computed from `Request#url`, `i18n.locales`, `i18n.routingStrategy` and `i18n.defaultLocale`

If the routing strategy is `prefix-other-locales`, for any URL that doesn't contain a locale, we return the `defaultLocale`.

This PR adds a bunch of data that we pass down the rendering of the page/endpoint, very similar to the same work I did when implementing `preferredLocale`.

Closes PLT-1207

## Testing

I added new test cases that should cover new use cases.

## Docs

https://github.com/withastro/docs/pull/5410

 /cc @withastro/maintainers-docs for feedback!

### CODE REVIEW

File: packages/astro/src/core/endpoint/index.ts
Comment (natemoo-re) [inline packages/astro/src/core/endpoint/index.ts:106]: I believe this could be reduced to the following?

```suggestion
				return currentLocale;
			}
```

**Review by natemoo-re — APPROVED** [review]
Code LGTM! Excited to see this API added!

File: packages/astro/src/@types/astro.ts
Comment (sarah11918) [inline packages/astro/src/@types/astro.ts:2119]: Just checking that there will eventually be a corresponding docs entry in this file for this, too, like the other ones currently listed in configuration reference?

File: .changeset/quick-toes-peel.md
Comment (sarah11918) [inline .changeset/quick-toes-peel.md:2]: Also checking that this is a patch, so only waiting for the feature to merge but not for the next minor release? If so, please help me remember when the corresponding docs PR should merge, since it's not on the normal release schedule!  :smile:  https://github.com/withastro/docs/pull/5410

File: packages/astro/src/@types/astro.ts
Comment (ematipico) [reply packages/astro/src/@types/astro.ts:2119]: I was hoping that this would be auto-generated, but it's not. I am not **when** we should update this https://docs.astro.build/en/reference/api-reference/#astro-global, but I suppose we should do it right before going out of experimental? 

Please let me know, and I will create a task for me to write the documentation

File: .changeset/quick-toes-peel.md
Comment (ematipico) [reply .changeset/quick-toes-peel.md:2]: Once the PR is merged, this new thing will go out in the **next** release (regardless of its entity). 

> remember when the corresponding docs PR should merge

Ideally, we should merge it before the next release. Although we don't have a schedule for that, I think it's also okay if the docs PR is released 1/2 day after.

File: packages/astro/src/@types/astro.ts
Comment (sarah11918) [reply packages/astro/src/@types/astro.ts:2119]: Ah, great point! 

To answer your question, No, these don't get auto-generated.

But, you made me realize that yes, this is not a **configuration** setting, and yes, should go on the `api-reference` instead, so no `@docs` are actually needed here in this file.

And yes, I agree that we'll keep it OFF the api reference page while it's still experimental, so nothing else required. That was my mistake!

**Review by bluwy — APPROVED** [review]

**Pull request merged**
